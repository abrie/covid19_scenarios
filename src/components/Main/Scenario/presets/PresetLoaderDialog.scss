@import '../../../../styles/variables';

.preset-loader-dialog-table {
  width: 100%;
}

.preset-loader-dialog-table-body {
  display: block;
  width: 100%;
  height: calc(100vh - 400px);
  overflow-y: auto;
  -webkit-overflow-scrolling: touch;
}

.preset-loader-dialog-row {
  display: flex;
  width: 100%;
}

.preset-loader-dialog-row-label {
  flex: 1;
  line-height: 2.5em;
}

@media screen and (max-width: 992px) {
  .preset-loader-dialog {
    .modal-content {
      top: 0;
      right: 0;
      bottom: 0;
      left: 0;
      display: fixed;
      max-height: 100vh;
    }
  }
}
