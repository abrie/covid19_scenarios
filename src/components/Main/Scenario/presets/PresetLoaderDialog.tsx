import React, { useEffect, useState, KeyboardEvent } from 'react'
import { useTranslation } from 'react-i18next'
import { useDebouncedCallback } from 'use-debounce'
import {
  Button,
  Modal,
  ModalBody,
  ModalFooter,
  ModalHeader,
  Table,
  Input,
  InputGroup,
  InputGroupAddon,
} from 'reactstrap'
import LoadPresetDialogRow, { LoadPresetDialogRecordProps } from './PresetLoaderDialogRow'
import './PresetLoaderDialog.scss'

const DEBOUNCE_DELAY = 500

export interface LoadPresetDialogProps {
  visible: boolean
  data: LoadPresetDialogRecordProps[]
  onLoadButtonClick: (id: string) => void
  onClose: () => void
}

const LoadPresetDialog = ({ data, visible, onLoadButtonClick, onClose }: LoadPresetDialogProps) => {
  const { t } = useTranslation()
  const [searchTerm, setSearchTerm] = useState('')
  const [filteredRows, setFilteredRows] = useState(data)
  const rows = filteredRows.map((item) => (
    <LoadPresetDialogRow key={item.value} data={item} onLoadButtonClick={onLoadButtonClick} />
  ))
  const executeFilter = (term: string) => {
    const hasSearchTerm = term.length > 0
    const filtered = hasSearchTerm ? data.filter((item) => item.label.toLowerCase().includes(term.toLowerCase())) : data

    setFilteredRows(filtered)
  }
  const [executeFilterDebounced] = useDebouncedCallback(executeFilter, DEBOUNCE_DELAY)

  useEffect(() => {
    if (visible) {
      /* TODO for some reason the useRef way doesn't work */
      setTimeout(() => {
        const element: HTMLInputElement | null = document.getElementById(
          'preset-loader-dialog-input',
        ) as HTMLInputElement

        if (element) {
          element.focus()
          element.select()
        }
      }, 0)
    }
  }, [visible])

  const onChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const { value } = event.target

    setSearchTerm(value)
    executeFilterDebounced(value)
  }

  const onKeyDown = (event: KeyboardEvent<HTMLInputElement>) => {
    const isEnterKey = event.keyCode === 13

    if (isEnterKey) {
      executeFilter(searchTerm)
    }
  }

  // TODO find out how to translate "Showing {0} out of {1} entries"
  return (
    <Modal
      className="height-fit preset-loader-dialog"
      centered
      size="lg"
      isOpen={visible}
      toggle={onClose}
      data-testid="PresetLoaderDialog"
    >
      <ModalHeader toggle={onClose}>{t('Change scenario')}</ModalHeader>
      <ModalBody>
        <InputGroup>
          <Input
            id="preset-loader-dialog-input"
            data-testid="PresetLoaderDialogInput"
            placeholder={t('Search')}
            onChange={onChange}
            value={searchTerm}
            autoComplete="off"
            onKeyDown={onKeyDown}
          />
          <InputGroupAddon addonType="append">
            <Button
              color="secondary"
              data-testid="PresetLoaderDialogClearButton"
              disabled={searchTerm === ''}
              onClick={() => {
                setSearchTerm('')
                executeFilter('')
              }}
            >
              {t('Clear')}
            </Button>
          </InputGroupAddon>
        </InputGroup>
        <Table className="preset-loader-dialog-table">
          <thead>
            <tr>
              <th>
                {t('Showing')} {filteredRows.length} {t('of')} {data.length} {t('entries')}
              </th>
            </tr>
          </thead>
          <tbody className="preset-loader-dialog-table-body">{rows}</tbody>
        </Table>
      </ModalBody>
      <ModalFooter>
        <Button color="primary" onClick={onClose} data-testid="PresetLoaderDialogCloseButton">
          {t('Done')}
        </Button>
      </ModalFooter>
    </Modal>
  )
}

export default LoadPresetDialog
