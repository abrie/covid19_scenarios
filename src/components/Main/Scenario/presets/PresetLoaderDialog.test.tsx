import React from 'react'
import { cleanup, fireEvent, render, RenderResult, wait } from '@testing-library/react'
import { noop } from 'lodash'
import PresetLoaderDialog from './PresetLoaderDialog'

const DATA = [
  {
    label: 'First row',
    value: '1',
  },
  {
    label: 'Second row',
    value: '2',
  },
  {
    label: 'Third row',
    value: '3',
  },
  {
    label: 'ABC',
    value: '4',
  },
]

describe('PresetLoaderDialog', () => {
  let onLoadButtonClick: (id: string) => void
  let onClose: () => void
  let wrapper: RenderResult

  beforeEach(() => {
    console.warn = noop
    console.error = noop

    onLoadButtonClick = jest.fn()
    onClose = jest.fn()
    wrapper = render(<PresetLoaderDialog data={DATA} onLoadButtonClick={onLoadButtonClick} onClose={onClose} visible />)
  })

  afterEach(cleanup)

  it('matches snapshot', () => {
    expect(wrapper.asFragment()).toMatchSnapshot()
  })

  it('renders title', () => {
    expect(wrapper.getByText('Change scenario')).not.toBeNull()
  })

  it('displays scenarios', () => {
    const loadButtons = wrapper.getAllByTestId('PresetLoaderDialogRowButton')

    expect(wrapper.getByText('Showing 4 of 4 entries')).not.toBeNull()
    expect(loadButtons).toHaveLength(4)
  })

  it('filters scenarios', async () => {
    const input = wrapper.getByTestId('PresetLoaderDialogInput')
    let loadButtons

    fireEvent.change(input, { target: { value: 'ABC' } })
    fireEvent.keyDown(input, { key: 'Enter', code: 'Enter' })

    await wait(() => wrapper.getByText('Showing 1 of 4 entries'))

    loadButtons = wrapper.getAllByTestId('PresetLoaderDialogRowButton')
    expect(loadButtons).toHaveLength(1)

    const clearButton = wrapper.getByTestId('PresetLoaderDialogClearButton')
    fireEvent.click(clearButton)

    await wait(() => wrapper.getByText('Showing 4 of 4 entries'))

    loadButtons = wrapper.getAllByTestId('PresetLoaderDialogRowButton')

    expect(loadButtons).toHaveLength(4)
  })

  describe('row button click triggers onLoadButtonClick handler', () => {
    it('when results not filtered', () => {
      const loadButton = wrapper.getAllByTestId('PresetLoaderDialogRowButton')

      fireEvent.click(loadButton[0])

      expect(onLoadButtonClick).toHaveBeenCalledWith('1')

      fireEvent.click(loadButton[1])

      expect(onLoadButtonClick).toHaveBeenCalledWith('2')
    })

    it('when results filtered', async () => {
      const input = wrapper.getByTestId('PresetLoaderDialogInput')

      fireEvent.change(input, { target: { value: 'ABC' } })
      fireEvent.keyDown(input, { key: 'Enter', code: 'Enter' })

      await wait(() => wrapper.getByText('Showing 1 of 4 entries'))

      const loadButtons = wrapper.getAllByTestId('PresetLoaderDialogRowButton')
      fireEvent.click(loadButtons[0])

      expect(onLoadButtonClick).toHaveBeenCalledWith('4')
    })
  })

  describe('row link click triggers onLoadButtonClick handler', () => {
    it('when results not filtered', () => {
      const loadButton = wrapper.getAllByTestId('PresetLoaderDialogRowLink')

      fireEvent.click(loadButton[0])

      expect(onLoadButtonClick).toHaveBeenCalledWith('1')

      fireEvent.click(loadButton[1])

      expect(onLoadButtonClick).toHaveBeenCalledWith('2')
    })

    it('when results filtered', async () => {
      const input = wrapper.getByTestId('PresetLoaderDialogInput')

      fireEvent.change(input, { target: { value: 'ABC' } })
      fireEvent.keyDown(input, { key: 'Enter', code: 'Enter' })

      await wait(() => wrapper.getByText('Showing 1 of 4 entries'))

      const loadButtons = wrapper.getAllByTestId('PresetLoaderDialogRowLink')
      fireEvent.click(loadButtons[0])

      expect(onLoadButtonClick).toHaveBeenCalledWith('4')
    })
  })

  it('close button triggers the onClose handler', () => {
    const closeButton = wrapper.getByTestId('PresetLoaderDialogCloseButton')
    fireEvent.click(closeButton)

    expect(onClose).toHaveBeenCalled()
  })
})
