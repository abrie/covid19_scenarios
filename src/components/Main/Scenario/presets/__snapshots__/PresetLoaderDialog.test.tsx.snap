// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PresetLoaderDialog matches snapshot 1`] = `<DocumentFragment />`;
